### PR TITLE
Add `/dstr` local Claude skill for data science test design review

### LINKED ISSUE #13366: `/dstr` Claude skill for Data Science Test Design Review
## Problem

Tests catch correctness but miss DS-specific failures (_e.g.,_ numeric edges, stale state after restart, data loss through copy/paste, pane desync) because these aren't intuitive without DS context.

## Solution

`/dstr` -- an opt-in tool that surfaces 1-3 concrete DS test design gaps on PRs.

- **GitHub Action**: reviewer comments `/dstr` on a PR during review
- **Local skill**: dev runs `/dstr` in Claude Code before pushing

## Behavior

- Analyzes PR diff, linked issues, and discussion through a DS lens
- Outputs specific assertions or test inputs (not vague advice)
- Stays silent (~87%) when not relevant
- Never runs automatically, never blocks CI

### PR DESCRIPTION
**Pull request by rodrigosf672**
## Summary

Adds `/dstr` — an opt-in tool that surfaces data science test design gaps on PRs. Available as both a GitHub Action (for reviewers) and a local Claude Code skill (for authors).

## Who is this for?

| Persona | How they use it | When |
|---------|----------------|------|
| **Reviewer/QA** | Comment `/dstr` on a PR during review | During review, before approval |
| **Dev (self-check)** | Run `/dstr` locally in Claude Code against their branch | Before pushing, while writing tests |

## What it does

Analyzes a PR's diff, linked issues, and discussion through a DS lens. Outputs 1-3 concrete test assertions the author likely wouldn't think of without DS background — or stays silent if nothing applies.

## Examples

For real comprehensive list of examples, see https://github.com/posit-dev/positron/issues/13404.

<details>
<summary><b>GitHub Action: reviewer comments <code>/dstr</code> on a DS-relevant PR</b></summary>

```
┌─────────────────────────────────────────────────────────────────────┐
│ PR #13245: Strip ANSI escapes when copying notebook output          │
├─────────────────────────────────────────────────────────────────────┤
│                                                                     │
│  rodrigosf672 commented:                                            │
│  > /dstr                                                            │
│                                                                     │
│  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ │
│                                                                     │
│  github-actions[bot] commented:                                     │
│  > 👀 Analyzing...                                                  │
│                                                                     │
│  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ │
│                                                                     │
│  github-actions[bot] updated comment:                               │
│                                                                     │
│  > **Test design gaps:**                                            │
│  >                                                                  │
│  > - **Copied numeric data survives ANSI stripping intact** —       │
│  >   Console output from `df.describe()` interleaves ANSI color     │
│  >   codes with numeric values (red for negatives). If the regex    │
│  >   is too greedy, it eats adjacent digits. Test by copying        │
│  >   output with ANSI codes adjacent to decimal points, then        │
│  >   assert `expect(clipboard).toBe('-3.14')` not `'-3.4'`.        │
│  >                                                                  │
│  > ---                                                              │
│  > 4,231 in / 87 out | Was this helpful? React 👍 or 👎            │
│                                                                     │
└─────────────────────────────────────────────────────────────────────┘
```

</details>

<details>
<summary><b>Local skill: dev runs <code>/dstr</code> in Claude Code before pushing</b></summary>

```
$ claude

> /dstr

  Reading branch diff against main... 3 files changed

  **Test design gaps:**

  - **Post-restart data explorer shows live data, not cached DOM** —
    Your test asserts `toBeVisible()` after kernel restart, but the
    real risk is a zombie view rendering stale rows. Assert
    `expect(table.getCell(0, 'id')).toBe('1')` after restart to
    verify the data round-tripped, not just that a container mounted.
```

</details>

<details>
<summary><b>Not relevant: tool stays quiet</b></summary>

```
$ claude

> /dstr

  Reading branch diff against main... 2 files changed (CSS, theme)

  Nothing stood out from a data science testing perspective here.
```

</details>

## Design principles

- **Opt-in only** — never runs automatically, never blocks CI
- **Aggressive relevance gate** — ~87% of PRs get a one-liner; only PRs touching data paths get bullets
- **Concrete, not vague** — every suggestion names a specific assertion or test input
- **Grounded in history** — calibrated against 5,000 merged PRs, 10 recurring failure patterns from real Positron bugs

## Safety

- Restricted to MEMBER/COLLABORATOR/OWNER (no external triggering)
- Pinned dependencies with lockfile (`npm ci`)
- Error messages sanitized (generic in PR, full details in workflow logs only)
- Whitespace-trimmed comment matching (follows `cla.yml` pattern)

## Design process

```mermaid
flowchart LR
    A[Literature Review] --> B[Heuristic Distillation] --> C[Skill Prototype] --> D[Action Build] --> E[Calibration] --> F[Pattern Extraction] --> G[Relevance Gate] --> H[Expert Panel] --> I[Prompt Hardening] --> J[Security Hardening] --> K[Dual Delivery]
```

| Step | Detail |
|------|--------|
| Literature Review | 21 DS textbooks, 2,400+ academic papers, 48 Posit enterprise customer stories |
| Heuristic Distillation | 86 heuristics across 13 domains |
| Skill Prototype | 8 iterations, layered output framework |
| Action Build | GitHub workflow + Node script + Claude API |
| Calibration | 5,000 merged PRs, subsystem classification |
| Pattern Extraction | 40+ bug-fix PRs deep-read, 10 recurring failure patterns |
| Relevance Gate | 87% silent, blocklist/allowlist, borderline case tuning |
| Expert Panel | DS, Testing/QA, DevEx, Reliability, DevRel review |
| Prompt Hardening | Weak oracle detection, concrete assertions required |
| Security Hardening | Permission gate, pinned deps, error sanitization |
| Dual Delivery | Action for reviewers + local skill for authors |

## Files

| File | Purpose |
|------|---------|
| `.github/workflows/dstr.yml` | Action: `issue_comment` trigger, permission gate, 1Password secrets |
| `.github/scripts/dstr/review.mjs` | Script: fetches PR context, calls Claude, posts comment |
| `.github/scripts/dstr/system-prompt.md` | Runtime prompt: relevance gate + 21 condensed heuristics |
| `.github/scripts/dstr/calibration.md` | Calibration: failure patterns, population stats, design decisions |
| `.github/scripts/dstr/heuristics-reference.md` | Full 86-heuristic library (13 domains) — the foundation |
| `.github/scripts/dstr/package.json` | Pinned dependencies |
| `.claude/commands/dstr.md` | Local skill: `/dstr` in Claude Code for pre-push self-check |

Addresses #13366

## Test plan

- [ ] Comment `/dstr` on a DS-relevant PR (Data Explorer, Variables Pane)
- [ ] Comment `/dstr` on a non-DS PR (CSS, config) — expect one-liner
- [ ] Run `/dstr` locally in Claude Code against current branch
- [ ] Verify permission gate blocks non-collaborator trigger
- [ ] Collect dev feedback on output quality via thumbs up/down reactions

### CODE REVIEW

**Comment by midleman** [comment]
I’m not totally clear on how this is meant to be used yet... Who is this for: the person opening the PR, a reviewer, or QA? And when would it come into play, before review, during, or after CI runs? Is this more of a self-check, or something reviewers are expected to rely on?

Before rolling this out more broadly, a few things feel worth thinking through. The first few are less about the prompt itself and more about how this runs behind the scenes, mainly around cost and safety:

1. Right now, anyone can trigger this just by commenting, even people outside the project. Since each run makes a paid API call, it might be safer to limit this to contributors or team members.
2. It only looks at the first 100 files in a PR. For larger PRs, that means it could miss things without making it obvious. It might help to either handle bigger PRs fully or at least call out when the review is incomplete.
3. Each time this runs, it pulls in the latest versions of its dependencies and executes them while our access tokens are available. If one of those packages were ever compromised upstream, that code would run with our credentials. To reduce that risk, we should lock dependency versions and make sure only those reviewed versions are used during each run.

**Comment by rodrigosf672** [comment]
> I’m not totally clear on how this is meant to be used yet... Who is this for: the person opening the PR, a reviewer, or QA? And when would it come into play, before review, during, or after CI runs? Is this more of a self-check, or something reviewers are expected to rely on?
> 
> Before rolling this out more broadly, a few things feel worth thinking through. The first few are less about the prompt itself and more about how this runs behind the scenes, mainly around cost and safety:
> 
> 1. Right now, anyone can trigger this just by commenting, even people outside the project. Since each run makes a paid API call, it might be safer to limit this to contributors or team members.
> 2. It only looks at the first 100 files in a PR. For larger PRs, that means it could miss things without making it obvious. It might help to either handle bigger PRs fully or at least call out when the review is incomplete.
> 3. Each time this runs, it pulls in the latest versions of its dependencies and executes them while our access tokens are available. If one of those packages were ever compromised upstream, that code would run with our credentials. To reduce that risk, we should lock dependency versions and make sure only those reviewed versions are used during each run.

@midleman Thank you, the latest commit addresses all the concerns:
1. Permission gate added: only MEMBER/COLLABORATOR/OWNER can trigger `/dstr`.
2. Full pagination: uses `octokit.paginate()`, no more 100-file cap.
3. Pinned dependencies added.

And about the audience and when they'd use it: 
1. It's opt-in for devs, reviewers or QA during PR reviews. Anyone (internal) can comment `/dstr` on PRs touching data paths to get a DS-focused perspective on test gaps. Not a gate, not automatic, and completely optional. 
2. Devs can also run `/dstr` locally as a Claude skill as a self-check before pushing. This way, there would be no need for them to get feedback _after_ pushing, hence potentially saving some time.

File: .github/workflows/dstr.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    types: [created]
+
+permissions:
+  contents: read
+  pull-requests: read
+  issues: write
+
+jobs:
+  dstr:
+    if: github.event.issue.pull_request && github.event.comment.body == '/dstr'
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/workflows/dstr.yml]: Fixed — added `author_association` check at job level restricting to MEMBER/COLLABORATOR/OWNER.

[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/workflows/dstr.yml]: Fixed — added a strip-whitespace step (same pattern as `cla.yml`) and gated all subsequent steps on the trimmed output matching `/dstr` exactly.

File: .github/workflows/dstr.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      - name: Load secret
+        uses: 1password/load-secrets-action@v4
+        with:
+          export-env: true
+        env:
+          OP_SERVICE_ACCOUNT_TOKEN: ${{ secrets.OP_SERVICE_ACCOUNT_TOKEN }}
+          ANTHROPIC_KEY: "op://Positron/Anthropic/credential"
+
+      - name: Install dependencies
+        run: npm install @octokit/rest @anthropic-ai/sdk
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/workflows/dstr.yml]: Fixed — added `package.json` + `package-lock.json` under `.github/scripts/dstr/` and switched to `npm ci --prefix .github/scripts/dstr`.

File: .github/scripts/dstr/review.mjs [hunk 3]
```diff
@@ -0,0 +1,261 @@
+import fs from "fs";
+import path from "path";
+import { Octokit } from "@octokit/rest";
+import Anthropic from "@anthropic-ai/sdk";
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Fixed — added copyright header, switched to `node:` prefixed imports, single quotes, and tabs. Matches `.github/scripts/release-screenshots/compare-and-report.mjs` style.

File: .github/scripts/dstr/review.mjs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+  // --- Fetch PR discussion comments (issue comments + review comments) ---
+  const prDiscussion = await fetchPRDiscussion();
+
+  // --- Fetch changed files ---
+  const { data: files } = await octokit.pulls.listFiles({
+    owner,
+    repo,
+    pull_number: prNumber,
+    per_page: 100,
+  });
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Fixed — now uses `octokit.paginate(octokit.pulls.listFiles, ...)` to fetch all files regardless of PR size.

File: .github/scripts/dstr/review.mjs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+  // --- Handle NOT_RELEVANT: brief acknowledgment ---
+  if (review === "NOT_RELEVANT" || review.includes("NOT_RELEVANT")) {
+    await octokit.issues.updateComment({
+      owner,
+      repo,
+      comment_id: placeholderComment.id,
+      body: `Nothing stood out from a data science testing perspective here.`,
+    });
+    process.exit(0);
+  }
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Fixed — now checks `review.split('\n')[0].trim() === 'NOT_RELEVANT'` (strict first-line equality instead of `includes`).

File: .github/scripts/dstr/review.mjs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+---
+<sub>${tokensUsed} | Was this helpful? React \u{1F44D} or \u{1F44E} to this comment</sub>`,
+  });
+} catch (error) {
+  await octokit.issues.updateComment({
+    owner,
+    repo,
+    comment_id: placeholderComment.id,
+    body: `:x: /dstr failed: ${error.message}`,
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Fixed — now normalizes with `error instanceof Error ? error.message : String(error)` before logging. Full error goes to stdout (workflow logs only); PR comment shows a generic message with a link to the logs.

File: .github/workflows/dstr.yml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+permissions:
+  contents: read
+  pull-requests: read
+  issues: write
+
+jobs:
+  dstr:
+    if: >
+      github.event.issue.pull_request &&
+      contains(fromJSON('["MEMBER","COLLABORATOR","OWNER"]'), github.event.comment.author_association)
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/workflows/dstr.yml]: Good point. Adding a concurrency block is reasonable for defense-in-depth, though in practice the permission gate + exact-match trigger makes accidental double-triggers unlikely. Will consider adding `concurrency: { group: dstr-${{ github.event.issue.number }}, cancel-in-progress: true }` if needed in the future.

File: .github/workflows/dstr.yml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            head -n 1 >> $GITHUB_OUTPUT
+
+      - uses: actions/checkout@v4
+        if: steps.strip_comment.outputs.body == '/dstr'
+
+      - uses: actions/setup-node@v4
+        if: steps.strip_comment.outputs.body == '/dstr'
+        with:
+          node-version: 22
+
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/workflows/dstr.yml]: Checked the repo -- `actions/checkout@v4` and `actions/setup-node@v4` are actually the current versions used across most workflows here (including `cla.yml`, `release-*.yml`). The `@v6` reference appears to be incorrect -- GitHub's latest major for checkout is v4 and setup-node is v4. No change needed.

File: .github/scripts/dstr/review.mjs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+			pull_number: prNumber,
+			per_page: 30,
+		});
+
+		for (const c of reviewComments) {
+			if (c.user?.type === 'Bot') continue;
+			const entry = `**${c.user?.login}** on \`${c.path}\`: ${c.body?.slice(0, 300) ?? ''}\n`;
+			if (totalLen + entry.length > MAX_COMMENTS_CHARS) break;
+			parts.push(entry);
+			totalLen += entry.length;
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Intentional trade-off. The 30-comment cap with a 5,000-char budget is deliberate -- we're feeding this into a Claude prompt with limited context. Active PRs with 30+ review comments would overflow the useful context window anyway. The most relevant discussion tends to be in the first comments and inline reviews. Paginating here would increase API calls without meaningfully improving output quality.

This is intentional by design and can be increased in the future if needed.

File: .github/scripts/dstr/review.mjs [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+const prNumber = event.issue.number;
+
+const { data: placeholderComment } = await octokit.issues.createComment({
+	owner,
+	repo,
+	issue_number: prNumber,
+	body: ':eyes: Analyzing...',
+});
+
+try {
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Fair observation, but intentionally scoped this way. If event parsing or client construction fails, the workflow fails visibly in the Actions tab with a stack trace in logs -- which is the right behavior for infrastructure errors. The try/catch is specifically for errors that occur *after* the placeholder comment exists, so we can update it rather than leave a dangling "Analyzing..." message. Wrapping everything would mean we'd need a nested try/catch for the cleanup itself.

I'll determine importance of this after testing on GitHub, and it'll be known whether such changes would be relevant.

File: .github/scripts/dstr/heuristics-reference.md [hunk 11]
```diff
@@ -0,0 +1,319 @@
+# Data Science Test Design — Full Heuristic Reference
+
+86 heuristics across 13 domains, distilled from 21 data science textbooks, 2,400+ academic papers, 48 enterprise customer deployments, and deep analysis of the Positron IDE codebase.
+
+The system prompt (`system-prompt.md`) uses a condensed 21-heuristic subset for the GitHub Action. This document is the full reference — used by the local `/ds-test-design` slash command and for maintaining/expanding the condensed set.
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/heuristics-reference.md]: This reference doc predates the PR -- it was written during the local skill development phase where the command was called `/ds-test-design`. The reference doc serves both: the local skill (`/dstr` in `.claude/commands/`) and the action's system prompt. Will update the naming reference for clarity.

File: .github/scripts/dstr/review.mjs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+	for (const file of files) {
+		const patch = file.patch ?? '[binary or too large]';
+		const truncatedPatch =
+			patch.length > MAX_FILE_PATCH_CHARS
+				? patch.slice(0, MAX_FILE_PATCH_CHARS) +
+					`\n... [truncated, ${patch.length - MAX_FILE_PATCH_CHARS} chars omitted]`
+				: patch;
+
+		const entry = `## ${file.filename}\nStatus: ${file.status} | +${file.additions} -${file.deletions}\n\n\`\`\`diff\n${truncatedPatch}\n\`\`\`\n`;
+
```
[the remark this answers is not in the document]
Comment (rodrigosf672) [reply .github/scripts/dstr/review.mjs]: Interesting edge case. In practice, triple backticks in a PR diff patch are rare (they'd need to appear in the actual changed lines of a .md or template literal file) and the model processes the full prompt as structured text, not as rendered Markdown. The risk of prompt injection via patch content is low given the model's instruction-following, but I could switch to a different fence (e.g. ~~~diff or XML tags) if this becomes a concern. Low priority for now.

File: .github/workflows/dstr.yml (hunk 7 above)
Comment (juliasilge) [inline .github/workflows/dstr.yml]: I don't believe we have yet set up Claude to run via GH action in the public repo and we need to be very careful about doing so. Let's not combine a PR that works on a new skill/command and prompting with a big change like this. I believe internally at Posit we have some different preferred options for running Claude in this way; one example is the Publisher repo.

For now, let's focus just on the command and prompting.

**Review by juliasilge — COMMENTED** [review]
Thank you so much for working on this! In general, my experience is that these AI tools generate way too many tests, as well as tests that are not substantive or useful (for example, testing the values of an `enum` or similar silly things) so prompting to help us write actually useful tests would be really worthwhile.

I spot checked some of the PRs that I have some context on, and I think it would be good to iterate on some of this prompting to give more useful suggestions for tests. From my spot checking, this prompting seems to generate ideas for tests that are related to the _general feature_ a PR is focused on but not very related to the PR itself.

I'll walk through a couple here to point out what I mean:

For #13098 it said this:

- **File with ambiguous types triggers wrong column rendering** — Open a CSV where column 1 has values `["1", "2", "NA", "4"]`. Assert the Data Explorer infers this as character/string (not numeric), and that the statistics pane shows string-type stats (count, unique) rather than numeric stats (mean, std). DuckDB sampling infers numeric when NA proportion is low.
- **Large file (>1M rows) doesn't hang the UI** — Open a 2M-row parquet file from the editor. Assert that `GetDataValues` returns within the configured timeout and that the first page renders with correct row count in the status bar.

That PR is focused on opening an R or Python dataframe object from an editor, not a CSV or a parquet file. Those test suggestions might be relevant to the Data Explorer generally (and we do have tests for reading files like that) but aren't relevant to this PR. It wouldn't be appropriate to add tests like this to that PR.

For #13069 it said this:

- **Loaded .RData file populates Variables Pane with correct values** — Save `x <- data.frame(a=1:3)` to `.RData`, load from a path with spaces (`C:\Users\My Name\data.RData`). Assert the Variables Pane shows `x` with 3 rows and column `a` containing `[1, 2, 3]`. The path fix addresses whether the file loads at all, but the test should verify data integrity, not just that no error is thrown.

That PR is about fixing a Windows path issue and not about how the data gets loaded; the actual loading of `.RData` files doesn't use any code we've written at all, actually! We wouldn't want to add tests that test whether base R code is correct.

For #13201 it said this:

- **dropTop during active carriage-return overwrite preserves correct line** — Write `"progress: 50%"` then `CR + "progress: 75%"` (overwrite in progress), call `dropTop(0)` as no-op, then `LF + "done"`. Assert final lines are `["progress: 75%", "done"]`. If `dropTop(n)` fires mid-CR-sequence (as a scrollback timer would), verify the overwrite target isn't shifted.


This is very much not related to what that PR is about at all.

Overall, I would love for us to work on prompting to help us write actually useful tests. A really valuable focus area! So far, this prompting seems to be pushing the AI tool more in the direction of "not related to the current work". Do you think it would be better to encode more traditional testing strategies? I wonder if the prompting containing words like "non-obvious" is pushing this in a non-helpful direction.

**Comment by rodrigosf672** [comment]
I'm closing this PR due to other higher value tasks in notebooks and other areas. Additionally, I'm closing it because the latest version of the skill's output (https://github.com/posit-dev/positron/issues/13404#issuecomment-4581697070) is too silent and the cases it comes up with are usually **edge cases**. I don't think that would be helpful.
  
A few lessons:
- Brevity and relevance pulled against each other. Making the output concise (2-3 bullets max) forced the tool to pick what was "worth saying"... and it consistently picked edge cases over the straightforward scenarios that would actually help someone write better tests.
- Too silent can be worse than too noisy. When the relevance gate filtered a PR as not DS-critical, that silence reads as a green light. At the same time, too noisy can degrade trust. The balance between the two was difficult to grapple with. 
- The tool worked better as a personal thinking tool than as a team-facing bot. The same knowledge base that's useful when I interrogate it directly doesn't survive compression into an automated comment. At the same time, understanding the separation between what the tool picks up from specific vs. general knowledge can be challenging as well. 
- Finding the balance between derived academic heuristics vs. actual DS-related bugs can be challenging. And as new bugs emerge, we'd have to think of a way to keep up with the repo changing (and potentially having to update heuristics?).  
- Exploration: overall, I found that to be a good exploration of Claude skill-making. Definitely gonna take the lessons towards future tasks that are higher in demand and that offer higher ROI for the team.
